### PR TITLE
Feature/aql gather executors

### PR DESCRIPTION
**Pull request by mchacki**
Implement Gather Executors in by-line version
Replaced Unsorting with ID.
Reimplemented sorting gather.

Jenkins
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/3613/

The one red single server test cannot be related.

### CODE REVIEW

File: arangod/Aql/SortingGatherExecutor.cpp [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+        TRI_ASSERT(_inputRows[_dependencyToFetch].state == ExecutionState::DONE);
+        adjustNrDone(_dependencyToFetch, 4);
+      }
+      ++_dependencyToFetch;
+    }
+    _initialized = true;
+    if (_nrDone >= _numberDependencies) {
+      return {ExecutionState::DONE, NoStats{}};
+    }
+    _strategy->prepare(_inputRows);
```
Comment (goedderz) [inline arangod/Aql/SortingGatherExecutor.cpp]: Perhaps this block could be moved into an `init()` method.

File: arangod/Aql/SortingGatherExecutor.cpp [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+      }
+      if (!_inputRows[_dependencyToFetch].row) {
+        TRI_ASSERT(_inputRows[_dependencyToFetch].state == ExecutionState::DONE);
+        adjustNrDone(_dependencyToFetch, 2);
+        if (_nrDone >= _numberDependencies) {
+          return {ExecutionState::DONE, NoStats{}};
+        }
+      }
+    }
+  }
```
Comment (goedderz) [inline arangod/Aql/SortingGatherExecutor.cpp]: There are 4 DONE checks in the code above
```cpp
        if (_nrDone >= _numberDependencies) {
          return {ExecutionState::DONE, NoStats{}};
        }
```
which I think can all safely be removed if you add it once here instead.

File: arangod/Aql/SortingGatherExecutor.h [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+  std::pair<ExecutionState, Stats> produceRow(OutputAqlItemRow& output);
+
+  void adjustNrDone(size_t dependency, size_t hass);
+
+  inline size_t numberOfRowsInFlight() const {
+    // For every not-done dependency we have one row in the buffers.
+    // Initially _numberDependencies is == 0 and _nrDone == 0 as well.
+    // This is due to the fact that dependencies are built AFTER this node
+    // and the number is yet unknown.
+    return _numberDependencies - _nrDone;
```
[the remark this answers is not in the document]
Comment (mchacki) [reply arangod/Aql/SortingGatherExecutor.h]: In the "in between" case we have to be in WAITING states where the numberOfRows cannot be used anyways.

[the remark this answers is not in the document]
Comment (mchacki) [reply arangod/Aql/SortingGatherExecutor.h]: FYI _dependencyToFetch is only the indicator which dependency has to be asked for the next row and jumps randomly between 0 and _numberDependencies-1

**Pull request merged**